boards/risc-v/esp32c3-legacy/common/src/esp32c3_board_twai.c

```diff
@@ -24,78 +27,615 @@
+ */
+
+static int twai_attach(struct sja1000_config_s *config,
+                           sja1000_handler_t handler, void *arg);
+
+static int twai_detach(struct sja1000_config_s *config);
+
+static uint32_t twai_getreg(struct sja1000_dev_s *sja_priv,
+                               uint32_t reg);
+
+static void twai_putreg(struct sja1000_dev_s *sja_priv,
+                           uint32_t reg, uint32_t value);
+
+static void twai_modifyreg32(struct sja1000_dev_s *sja_priv,
+      uint32_t reg, uint32_t clearbits, uint32_t setbits);
+
+static int twai_init_gpio(FAR struct sja1000_config_s *config);
+
+static int board_twai_initialise(int devno);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* CAN hardware-dependent bit-timing constant
+ * Used for calculating and checking bit-timing parameters
+ */
+
+static const struct can_bittiming_const_s esp32c3_twai_bittiming_const =
+{
+  .tseg1_min = 1,
+  .tseg1_max = 16,
+  .tseg2_min = 1,
+  .tseg2_max = 8,
+  .sjw_max   = 3,
+  .brp_min   = 1,
+  .brp_max   = 64,
+  .brp_inc   = 1,
+};
+
+/* A reference to a structure of this type must be passed to the SJA1000
+ * driver.  This structure provides information about the configuration
+ * of the SJA1000 and provides some board-specific hooks.
+ *
+ * Memory for this structure is provided by the caller.  It is not copied
+ * by the driver and is presumed to persist while the driver is active. The
+ * memory must be writable because, under certain circumstances, the driver
+ * may modify frequency or X plate resistance values.
+ */
+
+static struct esp32c3_twai_config_s
+    g_esp32c3_twai_config[] =
+{
+#ifdef CONFIG_ESP32C3_TWAI0
+    {
+      .config =
+        {
+          .bittiming_const = &esp32c3_twai_bittiming_const,
+          .port     = 0,
+          .periph   = ESP32C3_PERIPH_TWAI,
+          .irq      = ESP32C3_IRQ_TWAI,
+          .cpuint   = -ENOMEM,
+          .bitrate  = CONFIG_CAN_SJA1000_BITRATE,
+          .samplep  = CONFIG_CAN_SJA1000_SAMPLEP,
+          .sjw      = CONFIG_CAN_SJA1000_SJW,
+#ifdef CONFIG_CAN_LOOPBACK
+          .loopback = true,
+#else
+          .loopback = false,
+#endif /* #ifdef CONFIG_CAN_LOOPBACK */
+          .triple_sample = CONFIG_CAN_SJA1000_SAM,
+          .attach        = twai_attach,
+          .detach        = twai_detach,
+        },
+    },
+#endif /* #ifdef CONFIG_ESP32C3_TWAI0 */
+#ifdef CONFIG_ESP32C3_TWAI1
+#  error "TODO - TWAI1 not yet defined"
+#endif /* #ifdef CONFIG_ESP32C3_TWAI1 */
+};
+
+static struct sja1000_dev_s g_twai_priv[] =
+{
+#ifdef CONFIG_ESP32C3_TWAI0
+    {
+      .getreg = twai_getreg,
+      .putreg = twai_putreg,
+      .modifyreg32 = twai_modifyreg32,
+      .config = &g_esp32c3_twai_config[0].config,
+      .base = DR_REG_TWAI_BASE,
+    },
+#endif /* #ifdef CONFIG_ESP32C3_TWAI0 */
+};
+
```

Comment:
I am not sure about layering there. I would suggest to leave chop specific
```
      .putreg = twai_putreg,
      .modifyreg32 = twai_modifyreg32,
      .config = &g_esp32c3_twai_config[0].config,
      .base = DR_REG_TWAI_BASE,
```
in the `struct sja1000_config_s` and copied these for speedup into private `struct sja1000_dev_s` during chip initialization. This way the BSP level does not need to know any details about actual driver private `struct sja1000_dev_s`. The arguments against this option would be if even priv structure can be kept const (in Flash), but I expect that it should be usable for the device state storage. And yes, there is some duplication and space waste due to copied base and pointers to reg functions in my prefered solution, but it provides speedup and HW access functions simpler.